In [13]:
import pandas as pd
import sqlite3
import os
import matplotlib.pyplot as plt

print("Current Folder:")
print(os.getcwd())

print("\nFiles and Folders:")
print(os.listdir())

df = pd.read_csv("/Users/artitakle/Documents/Cyber Security Attack Analysis/Data/Processed/cleaned_attacks.csv")

conn = sqlite3.connect("cyber_security.db")

df.to_sql(
    "cyber_attacks",
    conn,
    if_exists="replace",
    index=False
)

conn.close()

print("Database created successfully")

Current Folder:
/Users/artitakle/Documents/Cyber Security Attack Analysis/Notebooks

Files and Folders:
['.DS_Store', '02_data_cleaning.ipynb', '04_sql_analysis.ipynb', 'cyber_security.db', '01_data_loading.ipynb', '03_EDA_analysis.ipynb']
Database created successfully


In [2]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("cyber_security.db")

query = """
SELECT name
FROM sqlite_master
WHERE type='table';
"""

pd.read_sql(query, conn)

,name
0,cyber_attacks


In [3]:
query = """
SELECT COUNT(*)
FROM cyber_attacks;
"""

pd.read_sql(query, conn)

,COUNT(*)
0,40000


In [17]:
query = """
SELECT `Attack Type`,
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY `Attack Type`
ORDER BY total_attacks DESC;
"""

result = pd.read_sql(query, conn)

result

attack_freq = pd.read_sql(query, conn)
attack_freq

,Attack Type,total_attacks
0,DDoS,13428
1,Malware,13307
2,Intrusion,13265


In [ ]:
query = """
SELECT "Severity Level",
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY "Severity Level";
"""

severity = pd.read_sql(query, conn)
severity

,Severity Level,total_attacks
0,High,13382
1,Low,13183
2,Medium,13435


In [6]:
query = """
SELECT Protocol,
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY Protocol;
"""

protocol = pd.read_sql(query, conn)
protocol

,Protocol,total_attacks
0,ICMP,13429
1,TCP,13272
2,UDP,13299


In [7]:
query = """
SELECT "Network Segment",
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY "Network Segment";
"""

segments = pd.read_sql(query, conn)
segments

,Network Segment,total_attacks
0,Segment A,13273
1,Segment B,13319
2,Segment C,13408


In [3]:
query = """
SELECT "Source IP Address",
       COUNT(*) AS attack_count
FROM cyber_attacks
GROUP BY "Source IP Address"
ORDER BY attack_count DESC
LIMIT 10;
"""

In [5]:
query = """
SELECT "Destination IP Address",
       COUNT(*) AS target_count
FROM cyber_attacks
GROUP BY "Destination IP Address"
ORDER BY target_count DESC
LIMIT 10;
"""

In [7]:
query = """
SELECT "Severity Level",
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY "Severity Level";
"""

In [9]:
query = """
SELECT Protocol,
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY Protocol;
"""

In [10]:
query = """
SELECT "Network Segment",
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY "Network Segment";
"""

In [11]:
query = """
SELECT "Geo-location Data",
       COUNT(*) AS total_attacks
FROM cyber_attacks
GROUP BY "Geo-location Data"
ORDER BY total_attacks DESC
LIMIT 10;
"""

In [12]:
query = """
SELECT *
FROM cyber_attacks
WHERE "Severity Level" = 'High';
"""

In [15]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("cyber_security.db")

schema = pd.read_sql("""
PRAGMA table_info(cyber_attacks);
""", conn)

schema

,cid,name,type,notnull,dflt_value,pk
0,0,Timestamp,TEXT,0,None,0
1,1,Source IP Address,TEXT,0,None,0
2,2,Destination IP Address,TEXT,0,None,0
3,3,Source Port,INTEGER,0,None,0
4,4,Destination Port,INTEGER,0,None,0
5,5,Protocol,TEXT,0,None,0
6,6,Packet Length,INTEGER,0,None,0
7,7,Packet Type,TEXT,0,None,0
8,8,Traffic Type,TEXT,0,None,0
9,9,Payload Data,TEXT,0,None,0


 Key Findings

1. Attack Distribution is Nearly Balanced

* The dataset contains **40,000 cyber attack records**.
* Attack types are almost evenly distributed:

  * DDoS: **13,428 attacks**
  * Malware: **13,307 attacks**
  * Intrusion: **13,265 attacks**
* No single attack type dominates, indicating organizations must prepare for multiple threat categories.



2. High-Severity Threats Represent Significant Risk

* High Severity: **13,382 attacks**
* Medium Severity: **13,435 attacks**
* Low Severity: **13,183 attacks**
* High and medium severity attacks account for nearly two-thirds of total incidents, requiring continuous monitoring and rapid response mechanisms.



3. Network Traffic Uses Multiple Attack Protocols

* ICMP: **13,429 attacks**
* UDP: **13,299 attacks**
* TCP: **13,272 attacks**
* Attack traffic is evenly distributed across protocols, suggesting attackers leverage diverse communication channels to bypass security controls.



4. All Network Segments Are Equally Exposed

* Segment A: **13,273 attacks**
* Segment B: **13,319 attacks**
* Segment C: **13,408 attacks**
* Cyber threats affect all network segments at similar levels, indicating organization-wide security measures are necessary rather than focusing on a single segment.



5. Attack Sources Are Highly Distributed

* Top source IP addresses generated only one recorded attack each.
* No single IP address appears repeatedly as a major threat source.
* This suggests distributed attack behavior, making traditional IP-blocking strategies less effective.



6. Attack Targets Are Widely Distributed

* Destination IP analysis shows attacks are spread across multiple target systems.
* No single destination IP dominates attack volume.
* Security monitoring should cover the entire network infrastructure rather than a few critical assets.



7. Geographic Attack Sources Are Diverse

* Top locations account for only 13–16 attacks each.
* No clear geographic hotspot exists.
* Threats originate from multiple regions, emphasizing the need for global threat intelligence and monitoring.


8. Security Environment Shows Constant Threat Activity

* Balanced attack types, severity levels, protocols, and network segments indicate a continuously active threat environment.
* Organizations should adopt layered cybersecurity defenses including:

  * IDS/IPS monitoring
  * Firewall protection
  * Anomaly detection systems
  * Incident response procedures

Business Recommendations

1. Implement real-time threat monitoring across all network segments.
2. Prioritize detection of high and medium severity attacks.
3. Strengthen protocol-level monitoring for TCP, UDP, and ICMP traffic.
4. Deploy anomaly-based detection instead of relying solely on IP blocking.
5. Conduct regular security assessments to identify emerging vulnerabilities.
6. Maintain centralized logging and incident response mechanisms.

Conclusion

> Analyzed 40,000 cybersecurity events using SQL and Python to identify attack trends, threat severity patterns, protocol usage, network exposure, and geographically distributed attack behavior, enabling data-driven cybersecurity monitoring and risk mitigation strategies.
